### PR TITLE
[SYCL-MLIR] Implement 5 SYCL types

### PR DESCRIPTION
**Pull request by whitneywhtsang**
Implemented SYCL `TupleValueHolder`, `TupleCopyAssignableValueHolder`, `AssertHappened`, `bfloat16` and `sub_group` types.

clang generated example:
```
%"struct.sycl::_V1::detail::TupleValueHolder" = type { i32 }
%"struct.sycl::_V1::detail::TupleCopyAssignableValueHolder" = type { %"struct.sycl::_V1::detail::TupleValueHolder" }
%"struct.sycl::_V1::detail::AssertHappened" = type { i32, [257 x i8], [257 x i8], [129 x i8], i32, i64, i64, i64, i64, i64, i64 }
%"class.sycl::_V1::ext::oneapi::bfloat16" = type { i16 }
%"struct.sycl::_V1::ext::oneapi::sub_group" = type { i8 }
```

Signed-off-by: Tsang, Whitney <whitney.tsang@intel.com>

### CODE REVIEW

File: mlir-sycl/lib/Dialect/IR/SYCLOpsDialect.cpp [hunk 1]
```diff
@@ -93,8 +93,6 @@ class SYCLOpAsmInterface final : public mlir::OpAsmDialectInterface {
       return "drw";
     case mlir::sycl::MemoryAccessMode::Atomic:
       return "ato";
-    default:
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Dialect/IR/SYCLOpsDialect.cpp:96]: To fix compile warnings

File: mlir-sycl/lib/Dialect/IR/SYCLOpsDialect.cpp [hunk 2]
```diff
@@ -115,43 +113,40 @@ class SYCLOpAsmInterface final : public mlir::OpAsmDialectInterface {
       return "hi";
     case mlir::sycl::MemoryTargetMode::ImageArray:
       return "ia";
-    default:
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Dialect/IR/SYCLOpsDialect.cpp:118]: To fix compile warnings

File: mlir-sycl/lib/Dialect/IR/SYCLOpsDialect.cpp [hunk 3]
```diff
@@ -115,43 +113,40 @@ class SYCLOpAsmInterface final : public mlir::OpAsmDialectInterface {
...
-      llvm_unreachable("Unhandled kind");
     }
   }
 };
 
 mlir::OpAsmDialectInterface::AliasResult
 SYCLOpAsmInterface::getAlias(mlir::Type Type, llvm::raw_ostream &OS) const {
   return llvm::TypeSwitch<mlir::Type, AliasResult>(Type)
-      .Case<mlir::sycl::AccessorType>([&](auto Ty) {
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Dialect/IR/SYCLOpsDialect.cpp:127]: Reordered. Nothing removed, and added
```
      .Case<mlir::sycl::AssertHappenedType, mlir::sycl::BFloat16Type>(
          [&](auto Ty) {
            OS << "sycl_" << decltype(Ty)::getMnemonic() << "_";
            return AliasResult::FinalAlias;
          })
      .Case<mlir::sycl::GetScalarOpType, mlir::sycl::TupleValueHolderType>(
          [&](auto Ty) {
            OS << "sycl_" << decltype(Ty)::getMnemonic() << "_"
               << Ty.getDataType() << "_";
            return AliasResult::FinalAlias;
          })
      .Case<mlir::sycl::TupleCopyAssignableValueHolderType>([&](auto Ty) {
        OS << "sycl_" << decltype(Ty)::getMnemonic() << "_" << Ty.getDataType()
           << "_";
        return AliasResult::OverridableAlias;
      })
```

File: mlir-sycl/test/Conversion/SYCLToLLVM/sycl-types-to-llvm.mlir [hunk 4]
```diff
@@ -18,12 +18,17 @@
...
-// CHECK: llvm.func @test_get_scalar_op(%arg0: !llvm.[[GETSCALAROP:struct<"class.sycl::_V1::detail::GetScalarOp.*", \(i32\)>]])
-// CHECK: llvm.func @test_nd_item(%arg0: !llvm.[[ND_ITEM_1:struct<"class.sycl::_V1::nd_item.*", \(]][[ITEM_1_TRUE]][[ITEM_BASE_1_TRUE]][[RANGE_1]][[ARRAY_1]][[SUFFIX]], [[ID_1]][[ARRAY_1]][[SUFFIX]], [[ID_1]][[ARRAY_1]][[SUFFIX]][[SUFFIX]][[SUFFIX]], [[ITEM_1_FALSE]][[ITEM_BASE_1_FALSE]][[RANGE_1]][[ARRAY_1]][[SUFFIX]], [[ID_1]][[ARRAY_1]][[SUFFIX]][[SUFFIX]][[SUFFIX]], [[GROUP_1]][[RANGE_1]][[ARRAY_1]][[SUFFIX]], [[RANGE_1]][[ARRAY_1]][[SUFFIX]], [[RANGE_1]][[ARRAY_1]][[SUFFIX]], [[ID_1]][[ARRAY_1]][[SUFFIX]][[SUFFIX]][[SUFFIX]])
+// CHECK: llvm.func @test_group(%arg0: !llvm.[[GROUP:struct<"class.sycl::_V1::group.*", \(]][[RANGE_1]][[ARRAY_1]][[SUFFIX]], [[RANGE_1]][[ARRAY_1]][[SUFFIX]], [[RANGE_1]][[ARRAY_1]][[SUFFIX]], [[ID_1]][[ARRAY_1]][[SUFFIX]][[SUFFIX]])
+// CHECK: llvm.func @test_get_op(%arg0: !llvm.[[GET_OP:struct<"class.sycl::_V1::detail::GetOp", \(i8\)>]])
+// CHECK: llvm.func @test_get_scalar_op(%arg0: !llvm.[[GET_SCALAR_OP:struct<"class.sycl::_V1::detail::GetScalarOp.*", \(i32\)>]])
```
Comment (whitneywhtsang) [inline mlir-sycl/test/Conversion/SYCLToLLVM/sycl-types-to-llvm.mlir:23]: NFC, renamed the variables.

File: polygeist/tools/cgeist/Lib/CodeGenTypes.cc [hunk 5]
```diff
@@ -1337,14 +1337,17 @@ mlir::Type CodeGenTypes::getMLIRType(clang::QualType QT, bool *ImplicitRef,
     const auto *RD = RT->getAsRecordDecl();
     if (mlirclang::isNamespaceSYCL(RD->getEnclosingNamespaceContext())) {
       const auto TypeName = RD->getName();
-      if (TypeName == "range" || TypeName == "nd_range" ||
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CodeGenTypes.cc:1340]: Reordered by alphabetic order, and added the 5 types.

File: polygeist/tools/cgeist/Test/Verification/sycl/types.cpp [hunk 6]
```diff
@@ -136,3 +146,29 @@ SYCL_EXTERNAL void vec_3(sycl::vec<sycl::cl_int, 4> vec) {}
...
+
+// CHECK-LABEL: func.func @_Z8atomic_1N4sycl3_V16atomicIiLNS0_6access13address_spaceE1EEE(
+// CHECK:          %arg0: memref<?x!sycl_atomic_i32_1_> {llvm.align = 8 : i64, llvm.byval = !sycl_atomic_i32_1_, llvm.noundef})
+// CHECK-SAME: attributes {[[SPIR_FUNCCC]], [[LINKEXT]], [[PASSTHROUGH]]
+SYCL_EXTERNAL void atomic_1(sycl::atomic<int> atomic_int) {}
+
+// CHECK-LABEL: func.func @_Z8atomic_2N4sycl3_V16atomicIfLNS0_6access13address_spaceE3EEE(
+// CHECK:          %arg0: memref<?x!sycl_atomic_f32_3_> {llvm.align = 8 : i64, llvm.byval = !sycl_atomic_f32_3_, llvm.noundef})
+// CHECK-SAME: attributes {[[SPIR_FUNCCC]], [[LINKEXT]], [[PASSTHROUGH]]
+SYCL_EXTERNAL void atomic_2(sycl::atomic<float, sycl::access::address_space::local_space> atomic_float) {}
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Test/Verification/sycl/types.cpp:158]: NFC, moved from line line 123, no changed.

File: polygeist/tools/cgeist/Test/Verification/sycl/types.cpp [hunk 7]
```diff
@@ -16,10 +16,15 @@
...
 // CHECK-DAG: !sycl_nd_range_2_ = !sycl.nd_range<[2], (!sycl.range<2>, !sycl.range<2>, !sycl.id<2>)>
 // CHECK-DAG: !sycl_get_scalar_op_i32_ = !sycl.get_scalar_op<[i32], (i32)>
-// CHECK-DAG: !sycl_atomic_f32_3_ = !sycl.atomic<[f32,3], (memref<?xf32, 3>)>
-// CHECK-DAG: !sycl_atomic_i32_1_ = !sycl.atomic<[i32,1], (memref<?xi32, 1>)>
+// CHECK-DAG: !sycl_tuple_value_holder_i32_ = !sycl.tuple_value_holder<[i32], (i32)>
+// CHECK-DAG: [[TUPLE_COPY_ASSIGNABLE_VALUE_HOLDER_TRUE:!sycl_tuple_copy_assignable_value_holder_i32_.*]] = !sycl.tuple_copy_assignable_value_holder<[i32, true], (!sycl.tuple_value_holder<[i32], (i32)>)>
+// CHECK-DAG: [[TUPLE_COPY_ASSIGNABLE_VALUE_HOLDER_FALSE:!sycl_tuple_copy_assignable_value_holder_i32_.*]] = !sycl.tuple_copy_assignable_value_holder<[i32, false], (!sycl.tuple_value_holder<[i32], (i32)>)>
 // CHECK-DAG: !sycl_vec_f32_8_ = !sycl.vec<[f32, 8], (vector<8xf32>)>
 // CHECK-DAG: !sycl_vec_i32_4_ = !sycl.vec<[i32, 4], (vector<4xi32>)>
+// CHECK-DAG: !sycl_atomic_f32_3_ = !sycl.atomic<[f32,3], (memref<?xf32, 3>)>
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Test/Verification/sycl/types.cpp:24]: NFC, moved from line line 19, no changed.

File: mlir-sycl/test/Conversion/SYCLToLLVM/sycl-types-to-llvm.mlir [hunk 8]
```diff
@@ -18,12 +18,17 @@
...
-// CHECK: llvm.func @test_get_op(%arg0: !llvm.[[GETOP:struct<"class.sycl::_V1::detail::GetOp", \(i8\)>]])
-// CHECK: llvm.func @test_get_scalar_op(%arg0: !llvm.[[GETSCALAROP:struct<"class.sycl::_V1::detail::GetScalarOp.*", \(i32\)>]])
-// CHECK: llvm.func @test_nd_item(%arg0: !llvm.[[ND_ITEM_1:struct<"class.sycl::_V1::nd_item.*", \(]][[ITEM_1_TRUE]][[ITEM_BASE_1_TRUE]][[RANGE_1]][[ARRAY_1]][[SUFFIX]], [[ID_1]][[ARRAY_1]][[SUFFIX]], [[ID_1]][[ARRAY_1]][[SUFFIX]][[SUFFIX]][[SUFFIX]], [[ITEM_1_FALSE]][[ITEM_BASE_1_FALSE]][[RANGE_1]][[ARRAY_1]][[SUFFIX]], [[ID_1]][[ARRAY_1]][[SUFFIX]][[SUFFIX]][[SUFFIX]], [[GROUP_1]][[RANGE_1]][[ARRAY_1]][[SUFFIX]], [[RANGE_1]][[ARRAY_1]][[SUFFIX]], [[RANGE_1]][[ARRAY_1]][[SUFFIX]], [[ID_1]][[ARRAY_1]][[SUFFIX]][[SUFFIX]][[SUFFIX]])
+// CHECK: llvm.func @test_group(%arg0: !llvm.[[GROUP:struct<"class.sycl::_V1::group.*", \(]][[RANGE_1]][[ARRAY_1]][[SUFFIX]], [[RANGE_1]][[ARRAY_1]][[SUFFIX]], [[RANGE_1]][[ARRAY_1]][[SUFFIX]], [[ID_1]][[ARRAY_1]][[SUFFIX]][[SUFFIX]])
```
Comment (etiotto) [inline mlir-sycl/test/Conversion/SYCLToLLVM/sycl-types-to-llvm.mlir:21]: This file is getting very large and is hard to understand which of the CHECKS are for which function. Can we put the CHECKS near the code of the functions they are for? Should we also add splits between the tests?

File: mlir-sycl/lib/Dialect/IR/SYCLOpsDialect.cpp (hunk 1 above)
Comment (etiotto) [reply mlir-sycl/lib/Dialect/IR/SYCLOpsDialect.cpp:96]: But then if one enumerator is added the unreachable is not triggered? Might be able to suppress the compiler warning that shows up if all enumerators are covered (not sure exactly how right now).

**Review by etiotto — APPROVED** [review]
LGTM - some inline comments

File: mlir-sycl/lib/Dialect/IR/SYCLOpsDialect.cpp (hunk 1 above)
Comment (whitneywhtsang) [reply mlir-sycl/lib/Dialect/IR/SYCLOpsDialect.cpp]: This change is unrelated to this PR, undo.

File: mlir-sycl/test/Conversion/SYCLToLLVM/sycl-types-to-llvm.mlir (hunk 8 above)
Comment (whitneywhtsang) [reply mlir-sycl/test/Conversion/SYCLToLLVM/sycl-types-to-llvm.mlir:107]: done.

**Comment by etiotto** [comment]
The description could list the LLVM IR produced by clang for these SYCL types.

**Comment by whitneywhtsang** [comment]
> The description could list the LLVM IR produced by clang for these SYCL types.

done.

File: mlir-sycl/lib/Dialect/IR/SYCLOpsDialect.cpp [hunk 9]
```diff
@@ -124,34 +124,33 @@ class SYCLOpAsmInterface final : public mlir::OpAsmDialectInterface {
...
-        return AliasResult::FinalAlias;
-      })
-      .Case<mlir::sycl::ItemType, mlir::sycl::ItemBaseType>([&](auto Ty) {
-        OS << "sycl_" << decltype(Ty)::getMnemonic() << "_" << Ty.getDimension()
-           << "_";
-        return AliasResult::OverridableAlias;
-      })
+      .Case<mlir::sycl::AssertHappenedType, mlir::sycl::BFloat16Type>(
+          [&](auto Ty) {
+            OS << "sycl_" << decltype(Ty)::getMnemonic() << "_";
```
Comment (victor-eds) [inline mlir-sycl/lib/Dialect/IR/SYCLOpsDialect.cpp:129]: No need for trailing "_".

Comment (whitneywhtsang) [reply mlir-sycl/lib/Dialect/IR/SYCLOpsDialect.cpp:129]: There is a need for trailing "_", as there are mnemonic like `bfloat16`, which ends with number.

Comment (victor-eds) [reply mlir-sycl/lib/Dialect/IR/SYCLOpsDialect.cpp:129]: True!

**Review by victor-eds — APPROVED** [review]

**Pull request merged**